gbn/queue.go

```diff
@@ -112,28 +236,116 @@ func (q *queue) resend(cb func(packet *PacketData) error) error {
 	q.topMtx.RUnlock()
 
 	if base == top {
+		q.awaitingCatchUpMu.Unlock()
+
 		return nil
 	}
 
+	// Prepare the queue for awaiting the resend catch up.
+	q.awaitedACK = (q.cfg.s + top - 1) % q.cfg.s
+	q.awaitedNACK = top
+
+	q.log.Tracef("Set awaitedACK to %d & awaitedNACK to %d",
+		q.awaitedACK, q.awaitedNACK)
+
+	q.awaitingCatchUp = true
+
+	// Drain the caughtUpSignal channel, in case no proceedAfterTime
+	// func was executed after the last resend catch up.
+	select {
+	case <-q.caughtUpSignal:
+	default:
+	}
+
 	q.log.Tracef("Resending the queue")
 
 	for base != top {
 		packet := q.content[base]
 
-		if err := cb(packet); err != nil {
+		if err := q.cfg.sendPkt(packet); err != nil {
+			q.awaitingCatchUpMu.Unlock()
+
 			return err
 		}
-		base = (base + 1) % q.s
+
+		base = (base + 1) % q.cfg.s
 
 		q.log.Tracef("Resent %d", packet.Seq)
 	}
 
+	// We hold the awaitingCatchUpMu mutex for the duration of the resend to
+	// ensure that we don't process the delayed ACKs for the packets we are
+	// resending, during the resend. If that would happen, we would start
+	// the "proceedAfterTime" callback timeout while still resending
+	// packets. That could mean that the NACK that the resent packets will
+	// trigger, might be received after the timeout has passed. That would
+	// cause the resend loop to trigger once more.
+	q.awaitingCatchUpMu.Unlock()
+
+	// Then await until we know that both parties are in sync.
+	q.awaitCatchUp(resendTimeout)
+
 	return nil
 }
 
-// processACK processes an incoming ACK of a given sequence number.
-func (q *queue) processACK(seq uint8) bool {
+// awaitCatchUp awaits that we either receive the awaited ACK or NACK signal
+// before returning. If we don't receive the awaited ACK or NACK signal before
+// 3X the resend timeout, the function will also return.
+// See the docs for the resend function for more details on why we need to await
+// the awaited ACK or NACK signal.
+//
+//nolint:cyclop
+func (q *queue) awaitCatchUp(resendTimeout time.Duration) {
+	ticker := time.NewTimer(resendTimeout * awaitingTimeoutMultiplier)
+	defer ticker.Stop()
+
+	q.log.Tracef("Awaiting catchup after resending the queue")
+
+catchupLoop:
+	for {
+		select {
+		case <-q.quit:
+			return
+		case <-q.awaitedACKSignal:
+			q.log.Tracef("Got awaitedACKSignal")
+
+			break catchupLoop
+		case <-q.awaitedNACKSignal:
+			q.log.Tracef("Got awaitedNACKSignal")
+
+			break catchupLoop
+		case <-ticker.C:
+			q.log.Tracef("Timed out while awaiting catchup")
+
+			q.awaitingCatchUpMu.Lock()
+			q.awaitingCatchUp = false
+
+			// Drain both the ACK & NACK signal channels.
+			select {
+			case <-q.awaitedACKSignal:
+			default:
+			}
+
+			select {
+			case <-q.awaitedNACKSignal:
+			default:
+			}
+
+			q.awaitingCatchUpMu.Unlock()
+
+			break catchupLoop
+		default:
```

Comment:
If we have a default case in a `select`, it basically means this will spin in a loop very very quickly and eat up a lot of CPU. Since we have a timeout ticker I think we can just remove the `default` case here.

Reply:
Ah thanks, you're right :)!